app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt

```diff
@@ -1,165 +1,292 @@
+                functionCall: Expression.FunctionCall,
+                variable: Variable,
+            ) : TypeCheckingError(listOf(functionCall, variable)) {
+                override val errorMessage = "The variable is called."
+            }
+
+            class ParameterCall(
+                functionCall: Expression.FunctionCall,
+                parameter: Function.Parameter,
+            ) : TypeCheckingError(listOf(functionCall, parameter)) {
+                override val errorMessage = "The parameter of a function is called."
+            }
+
+            class ConditionalTypesMismatch(
+                conditional: Expression.Conditional,
+                typeWhenTrue: Type,
+                typeWhenFalse: Type,
+            ) : TypeCheckingError(listOf(conditional)) {
+                override val errorMessage = "The results of a conditional operator have distinct types: $typeWhenTrue and $typeWhenFalse"
+            }
+
+            class NonConstantExpression(
+                expression: Expression,
+            ) : TypeCheckingError(listOf(expression)) {
+                override val errorMessage = "The expression is not constant (expected constant)."
+            }
+
+            class InvalidType(
+                expression: Expression,
+                type: Type,
+                expectedType: Type,
+            ) : TypeCheckingError(listOf(expression)) {
+                override val errorMessage = "The type of the expression is $expectedType (expected $type)."
+            }
+
+            class MissingReturnStatement(
+                function: Function,
+            ) : TypeCheckingError(listOf(function)) {
+                override val errorMessage = "The function neither returns unit, not has a 'return' statement."
+            }
         }
 
-        data class BreakOutsideOfLoop(val loopBreak: Statement.LoopBreak) : ControlFlowDiagnostic() {
+        sealed class VariablePropertiesError(astNodes: List<AstNode>) : ResolutionError(astNodes) {
             override fun isError() = true
-            override fun toString() = "Cannot use 'break' outside of loop"
+
+            class AssignmentToFunctionParameter(
+                parameter: Function.Parameter,
+                owner: Function,
+                assignedIn: Function
+            ) : VariablePropertiesError(listOf(owner, parameter, assignedIn)) {
+                override val errorMessage = "The parameter is owned by the function ${owner.name} and is assigned in the function ${assignedIn.name}."
+            }
         }
 
-        data class ContinuationOutsideOfLoop(val loopContinuation: Statement.LoopContinuation) : ControlFlowDiagnostic() {
-            override fun isError() = true
-            override fun toString() = "Cannot use 'continue' outside of loop"
+        sealed class ControlFlowDiagnostic(astNodes: List<AstNode>) : ResolutionError(astNodes) {
+
+            sealed class Warnings(astNodes: List<AstNode>) : ControlFlowDiagnostic(astNodes) {
```

Comment:
It's a little weird that a class representing warnings is derived from `ResolutionError` class.

Reply:
`ResolutionDiagnostic`?